File: Walya/Model/FsError.cs
namespace Walya.Model;

public class FsError
{
    public const string ERROR_DEFAULT = "Default";
    public const string ERROR_PARAMETER_COUNT_MISMATCH="TOO_FEW_PARAMETER";
    public const string ERROR_TYPE_MISMATCH = "TYPE_MISMATCH";
    public const string ERROR_TYPE_INVALID_PARAMETER = "TYPE_INVALID_PARAMETER";
    
    public string ErrorType { get; set; }
    public string ErrorMessage { get; set; }
    public object ErrorData { get; set; }

    public FsError(string messsage) : this(ERROR_DEFAULT, messsage, null)
    {
        
    }
    public FsError(string type,string messsage) : this(type, messsage, null)
    {
        
    }

    public FsError(string type, string message, string data)
    {
        this.ErrorType = type;
        this.ErrorMessage = message;
        this.ErrorData = data;
    }

    public override string ToString()
    {
        return $"{this.ErrorMessage} ({this.ErrorType})";
    }
}

File: js-port/funcscript-js/src/model/fs-error.js
class FsError {
  static ERROR_DEFAULT = 'Default';
  static ERROR_PARAMETER_COUNT_MISMATCH = 'TOO_FEW_PARAMETER';
  static ERROR_TYPE_MISMATCH = 'TYPE_MISMATCH';
  static ERROR_TYPE_INVALID_PARAMETER = 'TYPE_INVALID_PARAMETER';

  constructor(type, message, data = null) {
    this.errorType = type || FsError.ERROR_DEFAULT;
    this.errorMessage = message || '';
    this.errorData = data;
  }

  toString() {
    return `${this.errorMessage} (${this.errorType})`;
  }
}

module.exports = {
  FsError
};
